### PR TITLE
Add support for Arm64 PE files to System.Reflection.Metadata

### PR DESCRIPTION
**Pull request by davidwrighton**
- PE header DWORD is 0xAA64
- Arm64 PE files are 64 bit PE

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @nguerrera

**Comment by tmat** [comment]
@davidwrighton Thanks for adding this one. Would you mind adding some tests?

**Comment by davidwrighton** [comment]
@tmat I've added a set of tests that cover basic around multiple architectures per your request.

File: src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs [hunk 1]
```diff
@@ -59,10 +64,14 @@ private static unsafe void VerifyStrongNameSignatureDirectory(PEReader peReader,
             MethodDefinitionHandle entryPointHandle,
             Blob mvidFixup = default(Blob),
             byte[] privateKeyOpt = null,
-            bool publicSigned = false)
+            bool publicSigned = false,
+            Machine machine = 0)
         {
+            var peHeaderBuilder = new PEHeaderBuilder(imageCharacteristics : entryPointHandle.IsNil ? Characteristics.Dll : Characteristics.ExecutableImage,
```
Comment (tmat) [inline src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: nit: no space before ```:```

File: src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs [hunk 2]
```diff
@@ -320,8 +347,9 @@ private static MethodDefinitionHandle BasicValidationEmit(MetadataBuilder metada
             return mainMethodDef;
         }
 
-        [Fact]
-        public void Complex()
+        [Theory] // Do BasicValidation on common machine types
+        [MemberData(nameof(CommonMachineTypes))]
```
Comment (tmat) [inline src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: Why not simply

```
[Theory]
[InlineData(new[] { Machine.Unknown, ..., Machine.ArmThumb2 })]
```
?

File: src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs [hunk 3]
```diff
@@ -86,6 +95,23 @@ private static unsafe void VerifyStrongNameSignatureDirectory(PEReader peReader,
             peBlob.WriteContentTo(peStream);
         }
 
+        public static IEnumerable<object[]> AllMachineTypes()
```
Comment (tmat) [inline src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: ```=> Enum.GetValues(typeof(Machine)).Select(m => (object)m);```

Comment (tmat) [reply src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: Does it need to be weakly typed?

Comment (davidwrighton) [reply src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: Yes, and it actually need to be returned as a enumerator of object[]. The updated code is

return ((Machine[])Enum.GetValues(typeof(Machine))).Select(m => new object[]{(object)m});

File: src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs (hunk 2 above)
Comment (davidwrighton) [reply src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: I've just updated this to run for all machine types as well. It doesn't appear to be meaningfully slow.

File: src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs (hunk 1 above)
Comment (davidwrighton) [reply src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: Fixed

File: src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs [hunk 4]
```diff
@@ -95,21 +95,14 @@ private static unsafe void VerifyStrongNameSignatureDirectory(PEReader peReader,
             peBlob.WriteContentTo(peStream);
         }
 
-        public static IEnumerable<object[]> AllMachineTypes()
+        public static IEnumerable<object> AllMachineTypes()
         {
-            foreach (object enumValue in Enum.GetValues(typeof(Machine)))
+            return ((Machine[])Enum.GetValues(typeof(Machine))).Select(m => new object[]{(object)m});
+/*            foreach (object enumValue in Enum.GetValues(typeof(Machine)))
```
Comment (tmat) [inline src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: Forgotten comments?

Comment (davidwrighton) [reply src/System.Reflection.Metadata/tests/PortableExecutable/PEBuilderTests.cs]: Grumble... I had deleted those comments in my editor, but not actually saved them before committing. Sigh. Well, thanks for the signoff.

**Pull request merged**